datastore/sqlite_datastore.nim

```diff
@@ -339,8 +349,64 @@ method put*(
 
   return await self.put(key, data, timestamp())
 
-# method query*(
-#   self: SQLiteDatastore,
-#   query: ...): Future[?!(?...)] {.async, locks: "unknown".} =
-#
-#   return success ....some
+iterator query*(
+  self: SQLiteDatastore,
+  query: Query): Future[QueryResponse] =
+
+  let
+    queryStmt = QueryStmt.prepare(
+      self.env, queryStmtStr).expect("should not fail")
+
+    s = RawStmtPtr(queryStmt)
+
+  defer:
+    discard sqlite3_reset(s)
+    discard sqlite3_clear_bindings(s)
+    s.dispose
+
+  let
+    v = sqlite3_bind_text(s, 1.cint, query.key.id.cstring, -1.cint,
+      SQLITE_TRANSIENT_GCSAFE)
+
+  if not (v == SQLITE_OK):
+    raise (ref Defect)(msg: $sqlite3_errstr(v))
+
+  while true:
+    let
+      v = sqlite3_step(s)
+
+    case v
+    of SQLITE_ROW:
+      let
+        key = Key.init($sqlite3_column_text_not_null(
+          s, queryStmtIdCol)).expect("should not fail")
+
+        blob = sqlite3_column_blob(s, queryStmtDataCol)
+
+      # detect out-of-memory error
+      # see the conversion table and final paragraph of:
+      # https://www.sqlite.org/c3ref/column_blob.html
+      # see also https://www.sqlite.org/rescode.html
+
+      # the "data" column can be NULL so in order to detect an out-of-memory
+      # error it is necessary to check that the result is a null pointer and
+      # that the result code is an error code
+      if blob.isNil:
+        let
+          v = sqlite3_errcode(sqlite3_db_handle(s))
+
+        if not (v in [SQLITE_OK, SQLITE_ROW, SQLITE_DONE]):
+          raise (ref Defect)(msg: $sqlite3_errstr(v))
+
+      let
+        dataLen = sqlite3_column_bytes(s, queryStmtDataCol)
+        dataBytes = cast[ptr UncheckedArray[byte]](blob)
+        data = @(toOpenArray(dataBytes, 0, dataLen - 1))
+        fut = newFuture[QueryResponse]()
```

Comment:
will it work with blob==nil ? your testing code for queries doesn't include empty blobs

Reply:
I'll add some tests for that. Should already work since the approach is roughly the same for `method get`, which supports empty blobs.

Reply:
I've accounted for empty blobs in 32596eb.